### PR TITLE
vsrpc: Add usage telemetry

### PR DESCRIPTION
**Pull request by weikanglim**
Add usage telemetry for `vsrpc` server invocations.

To enable collection of telemetry, the server starts a goroutine that periodically ingests any newly emitted telemetry. On shutdown, any in-memory telemetry events is flushed to disk.

To collect usage telemetry, each server invocation emits a span for the duration of processing. The span contains RPC-specific attributes (ID -- for correlation, response error code, and method). The span also contains any usage attributes that were set throughout the server process lifetime (see `tracing.SetUsageAttribute` which is safe for multithreading). This works nicely with the slight imperfection that specific irrelevant attributes may persist across requests if not updated. We may readjust in the future, but largely attributes do end up being updated since the server currently acts more like an imperfect CLI wrapper after #3381.

### CODE REVIEW

File: cli/azd/cmd/middleware/telemetry.go [hunk 1]
```diff
@@ -89,171 +77,8 @@ func (m *TelemetryMiddleware) Run(ctx context.Context, next NextFn) (*actions.Ac
 	result.TraceID = span.SpanContext().TraceID().String()
 
 	if err != nil {
-		mapError(err, span)
+		cmd.MapError(err, span)
 	}
 
 	return result, err
 }
-
```
Comment (weikanglim) [inline cli/azd/cmd/middleware/telemetry.go:97]: This was a move-refactor to reshare mapping of application errors, see [commit](https://github.com/Azure/azure-dev/pull/3453/commits/7c4d5323af518c8318ac7384c32821c7c3d7033f). Also happy to split this into it's own PR if requested.

File: cli/azd/internal/vsrpc/server.go [hunk 2]
```diff
@@ -54,6 +61,19 @@ func (s *Server) Serve(l net.Listener) error {
...
+	ts := telemetry.GetTelemetrySystem()
+	go func() {
+		ticker := time.NewTicker(5 * time.Second)
+		for {
+			err := ts.RunBackgroundUpload(context.Background(), false)
+			if err != nil {
+				log.Printf("telemetry upload failed: %v", err)
+			}
+			<-ticker.C
+		}
```
Comment (wbreza) [inline cli/azd/internal/vsrpc/server.go:74]: Do we need something to stop this func from running forever? Perhaps leveraging go context with cancellation that can be checked for? I'm assuming otherwise it just runs until the process is terminated?

File: cli/azd/internal/vsrpc/server.go [hunk 3]
```diff
@@ -115,21 +135,47 @@ func serveRpc(w http.ResponseWriter, r *http.Request, handlers map[string]Handle
...
+
+			// Wrap the reply function to capture the response error returned by the handler before replying.
+			origReply := reply
+			reply = func(ctx context.Context, result interface{}, err error) error {
+				if err != nil {
+					respErr = err
+				}
+				return origReply(ctx, result, err)
+			}
+
```
Comment (wbreza) [inline cli/azd/internal/vsrpc/server.go:163]: Does the go server support any lifetime hooks / middleware support similar where we can register these type of per request cross cutting concerns?

**Review by wbreza — APPROVED** [review]
Added a few questions otherwise looks good.

File: cli/azd/internal/vsrpc/server.go (hunk 2 above)
Comment (weikanglim) [reply cli/azd/internal/vsrpc/server.go:74]: That's a great question.  The state that is being processed by the upload func is largely persisted in the filesystem and checkpointed enough by design. It is perfectly fine to perform a non-graceful shutdown and things would resume as "normal".

File: cli/azd/internal/vsrpc/server.go (hunk 3 above)
Comment (weikanglim) [reply cli/azd/internal/vsrpc/server.go:163]: The `Handler` interface would lend itself well to implementing composed functions (i.e., Middleware). For now, having a clear picture of the entire request handling flow may be more practical.

**Review by vhvb1989 — APPROVED** [review]

File: cli/azd/internal/vsrpc/server_service.go [hunk 4]
```diff
@@ -53,6 +55,14 @@ func (s *serverService) StopAsync(ctx context.Context) error {
 	// TODO(azure/azure-dev#3286): Need to think about how shutdown works. For now it is probably best to just have the
 	// client terminate `azd` once they know all outstanding RPCs have completed instead of trying to do a graceful
 	// shutdown on our end.
+
+	ts := telemetry.GetTelemetrySystem()
+	// Flush all in-memory telemetry data before stopping.
+	err := ts.Shutdown(ctx)
```
Comment (ellismg) [inline cli/azd/internal/vsrpc/server_service.go:61]: Do we need to arrange for the gorouitine that is doing the upload to stop?

**Review by ellismg — APPROVED** [review]
Cool to see this come alive!

File: cli/azd/internal/vsrpc/server_service.go (hunk 4 above)
Comment (weikanglim) [reply cli/azd/internal/vsrpc/server_service.go:61]: Added support for cancellation.

In general, we don't lose much if we abruptly shut down the process. The telemetry items are largely persisted on disk, and only removed after successful transmission. The worst-case scenario is an item is reuploaded if the shutdown occurred after an item is transmitted but the item isn't removed.

I was slightly hesitant of increasing complexity. Shutdowns are tricky to be proven to be time-bounded, and usually bad shutdown behaviors lead to unpredictability to the consuming tool, eventually resulting in a degraded experience for the user. But given that @wbreza also voiced a similar suggestion, I added the few lines of code to support this.

I manually tested the shutdown scenario by simulating an infinite loop within the telemetry process, and observed proper cancellation.
